100 refactor fitting for feature sets

ConnorPigg opened the pull request:
Closes #100, Closes #103, and closes #99 by refactoring fitting to allow fitting to a tree that contains multiple properties.

ConnorPigg committed:
Initial framework for multiproperty fitting.

Begins solution for #100 with an initial solution. Also adds
 various supporting features to ProfileProperty and PropertyDict.

ConnorPigg committed:
Testing for the interpolation and normalization.

Changes to the bayes testing in order to accomidate the change of
 the fitting to interpolate outside of the fitting process.

ConnorPigg committed:
Added tests for features used during fitting.

This includes a test for a new "feature_domain" property.
 This property is used to enable the testing of interpolated
 values prior to attempting to perform a fit.

ConnorPigg committed:
Updated bayes modelling to correct strategy.

ConnorPigg committed:
Update feature weighting to be based on errors.

This is done if errors are available. In either case weighting
 has a two norm of 1. Also updates modelling to not depend on
 properties having a length to use scalar properties.

ConnorPigg committed:
Test for bad fitting inputs.

Also allow PropertyDict.subset to take a single name since this is
 a potentially common occurance and if not handled correctly can
 provide strange results. Options were to throw an error or to
 correctly implement it.

ConnorPigg committed:
Change fitting procedure to expose models.

This allows users to configure or adjust parameters prior to
performing the fits. Adjusts the tests to match.

ConnorPigg committed:
Expose weighting option for fitting.

Also removed "unusable" guess function which would not work with
the composite modelling used during fitting.

ConnorPigg committed:
Cleaned up unused functions and docstrings.

jmborr commented:
@ConnorPigg please let me know when you think this PR is ready to be reviewed and I'll take a look

ConnorPigg committed:
Described available parameters for the model.

Also fixed the probability when there is only one depth.

ConnorPigg committed:
Format error messages correctly.

jmborr commented on idpflex/properties.py:
```diff
@@ -1185,33 +1228,74 @@ def __init__(self, name=None, qvalues=None, profile=None, errors=None):
+        if self.e is None or np.allclose(np.zeros(len(self.y)), self.e):
+            return np.ones(len(self.profile)) / np.sqrt(len(self.profile))
+        ws = self.profile / self.errors
+        ws[~np.isfinite(ws)] = self.profile[~np.isfinite(ws)]
```
To me it makes more sense to assign a zero weight to those q-values having non-finite weights because it indicates a problem for that particular q-value. It's safer to ignore the contribution of these q-values when calculating distances between properties.

For instance, if at q-value with index 42 we have `self.y[42]=np.nan` and `self.e[42]=1`, then `ws[42]=np.nan` but line 1262 will reassign `np.nan` to `ws[42]`, right?

ConnorPigg replied:
I should've written something like `ws[all(self.errors == 0)] = self.profile[all(self.errors == 0)]`. The intention was to simply prevent the introduction of nan by the division rather than correct bad data. 
You raise a valid point though about handling invalid input. This can be done by setting weights to zero for these as you suggest but should the error checking/handling happen instead when setting the profile/error values?

jmborr replied:
non-finite values should still be allowed in the profile and associated errors. A masking array of type `Bool` would be useful as additional attribute of the profile, to indicate which values are good and which ones are bad. Either that or have the functions using profiles deal with non-finite values.

ConnorPigg replied:
I would propose a method that will return a new filtered property, it could take an optional mask as input but will by default filter out points with nan, inf, or error==0.
```python
def filter(self, mask=None):
    if mask is None:
        mask = np.isfinite(self.profile) & np.isfinite(self.qvalues) & np.isfinite(self.errors) & (self.errors != 0)
    return self.__class__(profile=self.profile[mask], qvalues=self.qvalues[mask], name=self.name
                                       errors=self.errors[mask], node=self.node)
```
Therefore it is simple for a user to filter their data but otherwise allows the typical propagation of invalid data in the other parts of the code.

jmborr replied:
I think this method you propose is a better solution than extending the data structure with an additional attribute :smile:

jmborr commented on idpflex/properties.py:
```diff
@@ -1185,33 +1228,74 @@ def __init__(self, name=None, qvalues=None, profile=None, errors=None):
+    @property
+    def error_interpolator(self):
+        r"""
+        Return an interpolator from the profile data.
```
profile errors

jmborr commented on idpflex/properties.py:
```diff
@@ -1185,33 +1228,74 @@ def __init__(self, name=None, qvalues=None, profile=None, errors=None):
+        r"""Replace data with interpolated values at the specified qvalues."""
+        self.profile = self.interpolator(qvalues)
+        self.errors = self.error_interpolator(qvalues)
+        self.qvalues = qvalues.copy()
```
can we add a keyword `inplace=False` to the signature? I'd rather have this function's default behavior to return a property with interpolated values, and only overwrite the data and return `self` when `inplace` is set to True.

jmborr commented on idpflex/properties.py:
```diff
@@ -1185,33 +1228,74 @@ def __init__(self, name=None, qvalues=None, profile=None, errors=None):
-    r"""Mixin class providing a set of methods to load SANS data into a
-    profile property
-    """
+    r"""Mixin class providing a set of methods to load SANS data into a profile property."""  # noqa: E501
```
Is there a reason why you don't wrap the docstring and thus avoid flake8 to issue a complaint?

ConnorPigg replied:
The pydocstyle linting tool suggested this for readability I believe.

jmborr replied:
![Selection_016](https://user-images.githubusercontent.com/1136006/60030272-ad545700-9670-11e9-967b-8757d09a0401.png)

really? :smile:

ConnorPigg replied:
Both look the same for me with line wrapping :man_shrugging:
I can turn off/ignore pydocstyle if you do not like its conventions? :+1:

jmborr replied:
no, no, using `pydocstyle` is good

jmborr commented on idpflex/properties.py:
```diff
@@ -1420,7 +1501,7 @@ def from_crysol_fit(self, file_name):
                         args='-lm 20 -sm 0.6 -ns 500 -un 1 -eh -dro 0.075',
                         silent=True):
-        r"""Calculate profile with crysol from a PDB file
+        r"""Calculate profile with crysol from a PDB file.
```
:smile:

ConnorPigg replied:
The pydocstyle encourages periods in all of the docstrings so I've been adding them slowly. :smile:

jmborr reviewed (changes requested):
review of `properties.py`

jmborr commented on idpflex/bayes.py:
```diff
@@ -29,40 +32,87 @@ def __init__(self, xdata, ydata, interpolator_kind='linear',
+                                 'that does not match the xdata of the model. '
+                                 'Interpolate before or after.')
+            return amplitude * self._ydata
+        #     return amplitude * self._interpolator(x - center)
```
I think the `TabulatedFunctionModel` should be left untouched. What you want is a `ConstantVectorModel` where you return the same `y` values irrespective of the input `x`. You could define such a model.
That means functions `model_at_node` and `model_at_depth` have to be generalized with a keyword `fit_model` that defaults to  `ConstantVectorModel`. Also I'm thinking the signature of these models should be slightly changed so that we pass a reference to the property instead of components `x` and `y`. After all, `ConstantVectorModel` only requires `p.y` to be initialized.

jmborr reviewed (changes requested):
partial review of bayes.py

ConnorPigg committed:
Do not interpolate inplace by default.

Also addresses additional documenation changes for PR.

ConnorPigg committed:
Return TabulatedModel to interpolating.

ConnorPigg committed:
Use set_param_hint to be more lmfit idiomatic.

Also closes #102 by allowing global minimization method choice.

ConnorPigg committed:
Test changing fitting method.

This closes #102 by providing tests for using it.

ConnorPigg committed:
Test TabulatedModel with more inputs.

Previously rewrote tests to exclude interpolation so undoing that
change. Also, include testing the centering aspect of the model.

ConnorPigg committed:
Added filtering to ProfileProperty.

ConnorPigg committed:
Begins code structure for fixing #103.

However no testing has been completed and the "legacy"
multiproperty fitting has not been removed. Initial use with
`fitting_both.py` script indicates successful fitting.

ConnorPigg committed:
Fixes #103 by using new general modelling.

Replaces references to the MultiPropertyModel class for the new
model creation. Adjustments to the test to reflect the change in
model creation interface. Tests for acceptance of general models
and tests for parameter naming are still to be completed.

ConnorPigg committed:
Correct set checking for valid input.

ConnorPigg committed:
Fix #103 by adding tests specific to new changes.

This goes beyond the integration tests already present to test
specific expected behavior for the added general modelling.

ConnorPigg committed:
Update documentation regarding `models`.

This is to reflect that functions should not be directly passed
through to this parameter. It is expected that functions are
first converted to lmfit.Model classes or instances.

ConnorPigg committed:
Fixes bug of model objects not copying.

When Model instances are passed instead of classes, the object
was not properly handled which lead to each "use" of that model
template to overwrite the previous uses.

ConnorPigg committed:
Test "bad" inputs specific for the new modelling

ConnorPigg committed:
Rewriting code to try to be simpler to read.

ConnorPigg committed:
Change handling of variable argument type.

Attempt to be more "pythonic" by using try accept. Avoids using
`isinstance` to check types. Generalizes accepted types to single
object or iterables not just lists.

ConnorPigg committed:
Correct test assertion.

jmborr commented on idpflex/properties.py:
```diff
@@ -1171,47 +1214,126 @@ class ProfileProperty(object):
+    @property
+    def interpolator(self):
+        r"""
+        Return an interpolator from the profile data.
```
"a linear interpolator" :slightly_smiling_face: , unless you want to add an optional argument to pass as `kind` argument of `scipy.interpolate.interp1d`

jmborr commented on idpflex/properties.py:
```diff
@@ -1171,47 +1214,126 @@ class ProfileProperty(object):
+    @property
+    def error_interpolator(self):
+        r"""
+        Return an interpolator from the error data.
```
same as previous comment

jmborr commented on idpflex/properties.py:
```diff
@@ -1171,47 +1214,126 @@ class ProfileProperty(object):
+                                          fill_value='extrapolate')
+
+    def interpolate(self, qvalues, inplace=False):
+        r"""Replace data with interpolated values at the specified qvalues."""
```
maybe you should update this docstring to reflect the dual behavior of the method upon `inplace`

jmborr commented on idpflex/properties.py:
```diff
@@ -1171,47 +1214,126 @@ class ProfileProperty(object):
+            self.qvalues = qvalues.copy()
+            return self
+        # New instance of a property potentially using the subclass' init
+        return self.__class__(profile=self.interpolator(qvalues),
```
I would instead:
1. create a copy of `self`
2. Update attributes `profile`, `errors`, and `qvalues`
3. Return the copy

The reason is that if in the future we add one attribute to `ProfileProperty`, the call `self.__class__(profile=...node=self.node)` will miss setting the appropriate value of this new attribute.

jmborr reviewed (changes requested):
partial review

ConnorPigg committed:
Fix docs. Change `subset` `None` behavior.

PropertyDict no longer expects `None` for names to resturn self.
Instead, expect user to pass self.keys() for a complete subset.

jmborr reviewed (changes requested):
partial review of `bayes` module

ConnorPigg commented on idpflex/bayes.py:
```diff
@@ -1,68 +1,100 @@
+
+    def __init__(self, prop=None, **kwargs):
+        def constant_vector(x, scale, prop=None):
+            if not set(x).issuperset(prop.feature_domain):
```
The idea was it allows users to create custom models using functions that satisfy the requirement instead of custom classes. I thought this was simpler?
It allows
```
def func(x, ..., prop=None):
    return ...
CustomModel = lmfit.Model(func)
```
instead of
```
class CustomModel(lmfit.Model):
    def __init__(prop, ...):
        def func(x, ....):
            return ....
        super(CustomModel, self).__init__(func, ...)
```

jmborr replied:
If that's the case, then why does [ConstantVectorModel](https://github.com/jmborr/idpflex/blob/e1aa603dacf9146c363f510f74b8c293bb76c24e/idpflex/bayes.py#L46) requires an optional argument `prop` in the initialization method?

ConnorPigg commented on idpflex/bayes.py:
```diff
@@ -165,7 +197,198 @@ def fit_to_depth(tree, experiment, property_name, max_depth=5):
+    :class:`~lmfit.model.ModelResult`
+        Model for the depth
+    """
+    property_names = experiment.keys() if experiment is not None else None
```
The only reason I'd support passing scalar properties is that they can be used to influence the structure probabilities parameter if used with a model that does not have other parameters.
`def scalar_model(x, prop=None): return prop.y`

ConnorPigg replied:
How should the `property_type` be handled in combination with the names? Are the names simply filtered such that the only names used are those of the given property?

ConnorPigg commented on idpflex/bayes.py:
```diff
@@ -1,68 +1,100 @@
+                                 'with the domain of the profile being fitted.'
+                                 ' Interpolate before creating the model.')
+            return scale*prop.y
+        super(ConstantVectorModel, self).__init__(constant_vector, prop=prop,
```
I was just sticking with the convention in the code for super but this can be changed.

Passing `prop` to `Model` sets the keyword argument of `constant_vector` function during the fit evaluation function calls.

jmborr replied:
> I was just sticking with the convention in the code for super but this can be changed

This must be legacy code from the crazy time when I thought I had to support python 2.7 :stuck_out_tongue_closed_eyes:

> Passing prop to Model sets the keyword argument of constant_vector function during the fit evaluation function calls.

OK, some of `lmfit`'s dark magic

ConnorPigg commented on idpflex/bayes.py:
```diff
@@ -165,7 +197,198 @@ def fit_to_depth(tree, experiment, property_name, max_depth=5):
+        property_groups = [property_groups]
+
+    if len(property_groups) == 1:
+        return _create_model_from_property_group(property_groups[0], models)
```
If the user passes in a list of length 1 then the type error block will not be executed.

ConnorPigg commented on idpflex/properties.py:
```diff
@@ -1171,47 +1214,126 @@ class ProfileProperty(object):
+
+        Will keep the portion of the profile, qvalues, and errors that align
+        with true values in the mask. By default, the mask is true when all of
+        profile, qvalues, and errors are noninfinite and the errors is nonzero.
```
Then for clarity, I'd propose changing the name to something like `to_drop` or `to_keep` depending upon the desired behavior?

ConnorPigg commented on idpflex/bayes.py:
```diff
@@ -1,68 +1,100 @@
-        def tabulate(x, amplitude, center):
-            return amplitude * self._interpolator(x - center)
+        def tabulate(x, amplitude, center, prop=None):
+            return amplitude * prop.interpolator(x - center)
```
This was used to address copying model instances when attempting to use one model for many properties during a fit.
By using it in this way, I can use any instance a `lmfit.Model` as a template for multiple properties to be fit if I expect their function to take a keyword argument `prop` which can be changed for each property without having to know the attributes of the custom Model class.
This keyword argument changing is done [here](https://github.com/jmborr/idpflex/blob/e1aa603dacf9146c363f510f74b8c293bb76c24e/idpflex/bayes.py#L335-L337).

Perhaps the concern for performance can be addressed by attempting to optimize the property interpolation creation? I'd propose either a method (optionally "private" using _ prefix) of `create_interpolator`. This would preserve the simplicity of the model generation while addressing the performance concern (although I believe this way does incur more to lookup than using the model `._interpolator` directly).
```py
def create_interpolafortor(self, **interpolatorkws):
     self._interpolator = scipy.interpolate.interp1d(self.qvalues, self.profile, **interpolatorkws)

@property
def interpolator(self):
    if self._interpolator is None:
        self.create_interpolator()
    return self._interpolator
```

jmborr replied:
> This was used to address copying model instances when attempting to use one model for many properties during a fit.

Are you referring to statement `copy.deepcopy(m)` on [bayes.py:331](https://github.com/jmborr/idpflex/blob/e1aa603dacf9146c363f510f74b8c293bb76c24e/idpflex/bayes.py#L331)?
Since you allow to pass either the class or an instance of the class, there must be some use-case where you felt that passing the class did not solve the use-case. I can't see that so I'll have to stop by and you can explain to me.

> I'd propose either a method (optionally "private" using _ prefix) of create_interpolator

Yes, caching the interpolator will prevent multiple calls to the constructor. It is also a good idea having a dedicated function to create the interpolator, in case different options than the default ones ("linear", "extrapolate") are desired.

ConnorPigg committed:
Address PR change requests.

Adds optional filtering when taking a subset of PropertyDict.
Adds periods to documentation. Change calls of `super` to use no
arguments. Attempt to optimize property interpolation by only creating
the interpolator once instead of on each call.

ConnorPigg committed:
Add warnings. Consider logging instead/also.

ConnorPigg committed:
Expose subset filtering to model creation.

ConnorPigg committed:
Use random seed to reduce test flaky-ness.

This leads to reliability that failed tests are not due to just
randomness. This raises the issue that the occasional failures
from the randomness should be classified, fixed, and have
dedicated tests. Hypothesis tests may be useful for this.

ConnorPigg committed:
Persist interpolator kws through modifications.

Allow user settings for interpolators to persist through property
data modification by changing how interpolators are created. These
settings are now saved into an attribute of the property.

ConnorPigg committed:
Patched models to use modified _residual.

This is to enforce structure probabilities sum to 1.
Also rewrote global minimization testing to actually make sense.

ConnorPigg committed:
Change modelling to no longer bound probs.

Restructure modelling to focus on linear "mx + b" models. Change
to using proportions that are converted to probabilies for
interpretation. Remove tabulated only modeling and instead allow
choice of tabulated modelling in the restructured modelling
options.

ConnorPigg committed:
Test parameter naming conventions.

ConnorPigg committed:
Fix probability checking to use new names.

ConnorPigg committed:
Beware bounding total parameter.

ConnorPigg committed:
Update code and documentation of model creation.

Updates were to improve clarity.

ConnorPigg committed:
Fix bug in when to vary slope.

ConnorPigg committed:
Address overwriting parameter hints.

This involves being more verbose and adding tests for these
parameter hints. To simplify this, this changes the tabulated model
to use the same parameter names to simplify the logic to hard coded
parameter names.

ConnorPigg committed:
Fix bug where tabulated choice was ignored.

The `use_tabulated` was not being passed down from create to depth
to the building block functions.

ConnorPigg committed:
Distinguish between composite and simple tabulate.

The difference places restrictions on the composite to not allow
simultaneous centering and interpolation to different qvalues.
The restriction is in place because from the experimental feature
vector it is difficult to determine what section should be used
as the new qvalues. Therefore require user to interpolate before
and simply allow centering in composite.
